lookout/style/format/quality_report_noisy.py

```diff
@@ -262,59 +268,85 @@ def quality_report_noisy(true_repo: str, noisy_repo: str, bblfsh: str, language:
-    for i, (prec, rec) in enumerate(zip(precisions, recalls)):
-        if prec < precision_threshold:
-            break
-        confidence_threshold_exp = round(rules.rules[i].stats.conf, 3)
-        rec_threshold_prec = rec
-
-    # compile the precision-recall curve
-    path_to_figure = os.path.join(dir_output, "pr_curve_jquery.png")
-    plot_curve("jquery/jquery", numpy.asarray(recalls), numpy.asarray(precisions),
-               precision_threshold, rec_threshold_prec, confidence_threshold_exp, path_to_figure)
+    repositories = []
+    precisions, recalls = (defaultdict(list) for _ in range(2))
+    n_mistakes, rec_threshold_prec, prec_max_rec, confidence_threshold_exp, max_rec, \
+        n_rules, n_rules_filtered = ({} for _ in range(7))
+    for url in REPOSITORIES.splitlines():
+        repo = url.split("/")[-1]
+        log.info("Fetching %s", url)
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            git_dir = os.path.join(tmpdirname, repo)
+            git_dir_noisy = os.path.join(tmpdirname, repo + "_noisy")
+            cmd1 = "git clone --single-branch --branch master %s %s" % (url, git_dir)
+            cmd2 = "git clone --single-branch --branch style-noise-1-per-file %s %s" % (
+                    url, git_dir_noisy)
+            try:
+                subprocess.check_call(cmd1.split())
+                subprocess.check_call(cmd2.split())
+            except subprocess.CalledProcessError:
+                raise ConnectionError("Unable to fetch repository %s" % repo)
+            input_pattern = os.path.join(git_dir, "**", "*.js")
+            input_pattern_noisy = os.path.join(git_dir_noisy, "**", "*.js")
+            true_content = get_content_from_repo(input_pattern)
+            noisy_content = get_content_from_repo(input_pattern_noisy)
+
+            true_files, noisy_files, start_changes = get_difflib_changes(true_content,
+                                                                         noisy_content)
+            if not true_files:
+                raise ValueError("Noisy repo should count at least one artificial mistake")
+            log.info("Number of files modified by adding style noise: %d / %d", len(true_files),
+                     len(true_content))
+            del true_content, noisy_content
+
+            try:
+                client = BblfshClient(bblfsh)
+                model_path = os.path.join(git_dir_noisy, "style-analyzer-model", "model.asdf")
+                analyzer = FormatModel().load(model_path)
+                rules = analyzer[language]
+                feature_extractor = FeatureExtractor(language=language,
+                                                     **rules.origin_config["feature_extractor"])
+                vnodes_y_true = files2vnodes(true_files, feature_extractor, client)
+                mispreds_noise = files2mispreds(noisy_files, feature_extractor, rules, client, log)
+            finally:
+                client._channel.close()
```

Comment:
@warenlg Can you please submit `def BblfshClient.close()` implementation PR to bblfsh/bblfsh-client after the release so that we don't use protected API here.

Reply:
and also if we have an implementation for 
```python
with BblfshClient(bblfsh) as client:
    # do whatever
```
it will be super nice